src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java

```diff
@@ -0,0 +1,216 @@
+		}
+	}
+
+	// Run this on local players whenever their information is out of date
+	void refreshLocalPlayer(ProxiedPlayer player) {
+		MMLog.fine(() -> "Refreshing local player " + player.getName());
+		RemotePlayerBungee localPlayer = fromLocal(player, true);
+
+		// update local player with data
+		if (updatePlayerLocal(localPlayer, false)) {
+			localPlayer.broadcast();
+		}
+	}
+
+	void remotePlayerChange(JsonObject data) {
+		if (data == null) {
+			MMLog.severe(() -> "Null player data recieved from an unknown source!");
+			return;
+		}
+		RemotePlayerAbstraction player = RemotePlayerAbstraction.from(data);
+		if (player == null) {
+			// something really bad happened
+			MMLog.severe(() -> "Invalid player data recieved from an unknown source!");
+			return;
+		}
+
+		updatePlayerLocal(player, true);
+	}
+
+	boolean updatePlayerLocal(RemotePlayerAbstraction player, boolean isRemote) {
+		RemotePlayerData oldPlayerData = getRemotePlayer(player.mUuid);
+		String serverType = player.getServerType();
+		RemotePlayerAbstraction oldPlayer = oldPlayerData != null ? oldPlayerData.get(serverType) : null;
+
+		// Update the player before calling events
+		super.updatePlayer(player);
+
+		if (player.mIsOnline && (oldPlayer == null || !oldPlayer.mIsOnline)) {
+			RemotePlayerLoadedEventBungee remotePE = new RemotePlayerLoadedEventBungee(player);
+			ProxyServer.getInstance().getPluginManager().callEvent(remotePE);
+			MMLog.info(() -> "Loaded player: " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
+			return true;
+		}
+
+		@Nullable ProxiedPlayer localPlayer = ProxyServer.getInstance().getPlayer(player.mUuid);
+		if (isRemote && serverType.equals(RemotePlayerBungee.SERVER_TYPE) && localPlayer != null && localPlayer.isConnected()) {
+			// Player logged off on remote shard, but is locally online.
+			// This can happen if the remote shard was not notified the player logged in here in time.
+			MMLog.warning(() -> "Detected race condition, triggering refresh on " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
+			refreshLocalPlayer(localPlayer);
+			return false;
+		}
+
+		if (!player.mIsOnline && (oldPlayer == null || oldPlayer.mIsOnline)) {
+			MMLog.info(() -> "Unloaded player: " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
+			RemotePlayerUnloadedEventBungee remotePE = new RemotePlayerUnloadedEventBungee(player);
+			ProxyServer.getInstance().getPluginManager().callEvent(remotePE);
+			return true;
+		} else if (!isRemote || !player.isSimilar(oldPlayer)) {
+			RemotePlayerUpdatedEventBungee remotePE = new RemotePlayerUpdatedEventBungee(player);
+			ProxyServer.getInstance().getPluginManager().callEvent(remotePE);
+			MMLog.info(() -> "Updated player: " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
+			return true;
+		} else {
+			MMLog.warning(() -> "Ignored player: " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
```

Comment:
I don't think ignoring data should be a warning as a general rule - either there's nothing to update, and it can be `fine` or lower priority, or something went wrong trying to understand remote data, which should be a warning with more info.